### PR TITLE
Configurable Event Version in AuditSink

### PR DESCRIPTION
**Pull request by pbarker**
/kind api-change

Allows audit sinks to configure the version of events emitted https://github.com/kubernetes/kubernetes/issues/71464

```release-note
add event version configuration to audit sinks
```

### CODE REVIEW

**Comment by liggitt** [comment]
/hold

there's [work in progress](https://github.com/kubernetes/enhancements/pull/736) to move admission webhooks forward around the following things:
* specifying acceptable versions of received objects
* http for localhost destinations
* specifying a port for service reference destinations
* configurable timeout

the intent of the work coming out of that is to adjust all the webhookish things consistently

File: api/openapi-spec/swagger.json [hunk 1]
```diff
@@ -78637,6 +78637,10 @@
      "webhook"
     ],
     "properties": {
+     "eventVersion": {
+      "description": "EventVersion is the version of the event to emit. Default v1",
```
Comment (mourya007) [inline api/openapi-spec/swagger.json]: Just a query that right now it is handling only v1 vesion, So what happend if user specified v1alpha1 or else. Will it respond to the same? Do we need to handle for other versions? Thanks in advance

**Comment by pbarker** [comment]
@liggitt updated to conform with the new standard, let me know what you think
/assign @liggitt

File: pkg/apis/auditregistration/validation/validation.go [hunk 2]
```diff
@@ -48,6 +51,10 @@ func ValidateWebhook(w auditregistration.Webhook, fldPath *field.Path) field.Err
 		allErrs = append(allErrs, ValidateWebhookThrottleConfig(w.Throttle, fldPath.Child("throttle"))...)
 	}
 
+	for _, version := range w.EventVersions {
+		allErrs = append(allErrs, ValidateEventVersion(version, fldPath.Child("eventVersion"))...)
```
Comment (liggitt) [inline pkg/apis/auditregistration/validation/validation.go]: this requires all versions to be recognized by the apiserver, which is not quite what we want

1. this is called on update, which means a failure here could prevent update of an existing persisted object
2. requiring recognition of all items in the list means a sink couldn't register for `v1beta1,v1` against a server which only supports `v1beta1`. This scenario is main reason to accept multiple versions.

what we want is:
1. on create, require at least one recognized version
2. on update, if the existing object has at least one recognized version, require the new object to as well

then, when sending events to the audit sink, iterate through the accepted versions, sending the first one we know how to handle

Comment (pbarker) [reply pkg/apis/auditregistration/validation/validation.go]: updated
> 2. on update, if the existing object has at least one recognized version, require the new object to as well

I simply handled update the same as create, requiring a recognized version

Comment (liggitt) [reply pkg/apis/auditregistration/validation/validation.go]: > I simply handled update the same as create, requiring a recognized version

That's unsafe in HA skew and rollback scenarios where the stored data was persisted via a server that supported audit event versions the current server doesn't, and only claimed support for those versions

**Comment by xmudrii** [comment]
Hello! The code freeze is coming this Friday and I'd just like to check is this still planned to be merged in the next 10 days?

**Comment by xmudrii** [comment]
Prow seems stuck on this PR so trying to retest
/retest

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/73547#discussion_r260013653 is still outstanding

File: pkg/apis/auditregistration/validation/validation.go (hunk 2 above)
Comment (pbarker) [reply pkg/apis/auditregistration/validation/validation.go]: > on update, if the existing object has at least one recognized version, require the new object to as well
could you clarify what should happen if the old object does not have a supported version by a new server? Wouldn't we still want to verify the new object does, and replace the old one?

Comment (liggitt) [reply pkg/apis/auditregistration/validation/validation.go]: > could you clarify what should happen if the old object does not have a supported version by a new server? Wouldn't we still want to verify the new object does, and replace the old one?

if the existing persisted object does not have a supported version (for whatever reason), we should not require the new object to either. this allows unrelated updates (labeling, removing finalizers, updating CA bundle, etc)

Comment (liggitt) [reply pkg/apis/auditregistration/validation/validation.go]: the validation in https://github.com/kubernetes/kubernetes/pull/74998/commits/eb025089ff855de403f646aed1973bcd560fdc26#diff-54fdd67617b0f2ef1b5a09148f77303eR116 is a good example of what we're going for

**Comment by claurence** [comment]
Hi all, enhancements lead here - it seems like this PR isn't ready to merge and we've passed code freeze so I'm going to remove this from the 1.14 milestone.  If you feel this was done in error please talk with your SIG about getting an exception or come to the sig-release slack channel to discuss

**Comment by Pothulapati** [comment]
Hello! I'm part of the bug triage team for the 1.16 release cycle and considering this PR has not been updated for a long time, I'd like to check what's the status of this PR. The code freeze is starting 29th August (about 2.5 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time. 

As the PR is tagged for 1.16, is it still planned for this release?

**Comment by liggitt** [comment]
/close

clearing out PRs not being actively worked on. feel free to reopen when ready to resume work.

**Pull request closed without merging**